### PR TITLE
add sort to cart items

### PR DESCRIPTION
**Pull request by patrickweh**

### CODE REVIEW

File: src/Livewire/Portal/Shop/WatchlistCard.php [hunk 1]
```diff
@@ -22,7 +23,7 @@ class WatchlistCard extends Component
 
     public function mount(Cart $cart): void
     {
-        $cart->loadMissing('cartItems');
+        $cart->loadMissing(['cartItems' => fn (BelongsTo $query) => $query->ordered()]);
```
Comment (SirSplasch) [inline src/Livewire/Portal/Shop/WatchlistCard.php:26]: Mustn't this be HasMany?

File: src/Livewire/Cart/Watchlists.php [hunk 2]
```diff
@@ -33,7 +34,7 @@ public function getCarts(): Collection
                     ->orWhere('is_public', true);
             })
             ->with([
-                'cartItems',
+                'cartItems' => fn (HasMany $query) => $query->ordered(),
```
Comment (SirSplasch) [inline src/Livewire/Cart/Watchlists.php:37]: Why must this be done? Don't the cartItems have the newly added OrderedScope?

File: src/Livewire/Portal/Shop/WatchlistCard.php (hunk 1 above)
Comment (SirSplasch) [inline src/Livewire/Portal/Shop/WatchlistCard.php:26]: Why must this be done? Don't the cartItems have the newly added OrderedScope?

File: src/Rulesets/CartItem/CreateCartItemRuleset.php [hunk 3]
```diff
@@ -30,6 +30,10 @@ public function rules(): array
                 'integer',
                 app(ModelExists::class, ['model' => VatRate::class]),
             ],
+            'order_column' => [
```
Comment (SirSplasch) [inline src/Rulesets/CartItem/CreateCartItemRuleset.php:33]: negative integers allowed? I don't think so.

File: src/Rulesets/CartItem/UpdateCartItemRuleset.php [hunk 4]
```diff
@@ -19,6 +19,10 @@ public function rules(): array
                 'integer',
                 app(ModelExists::class, ['model' => CartItem::class]),
             ],
+            'order_column' => [
```
Comment (SirSplasch) [inline src/Rulesets/CartItem/UpdateCartItemRuleset.php:22]: negative integers allowed? I don't think so.

File: database/migrations/2024_10_30_090515_add_order_column_to_cart_items_table.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class() extends Migration
+{
+    public function up(): void
+    {
+        Schema::table('cart_items', function (Blueprint $table) {
+            $table->unsignedInteger('order_column')->after('vat_rate_id')->nullable()->index();
```
Comment (SirSplasch) [inline database/migrations/2024_10_30_090515_add_order_column_to_cart_items_table.php:12]: I would put the order_column after total. As it is only used for sorting.
If you decide to move the order_column, adjust the ordering in the corresponding rulesets as well.

File: src/Traits/SortableTrait.php [hunk 6]
```diff
@@ -2,18 +2,54 @@
...
+            if ($model->isDirty($orderColumn) && ! $model->getIsSorted()) {
+                $newPosition = $model->$orderColumn;
+                $model->$orderColumn = $model->getRawOriginal($orderColumn);
+
+                $model->moveToPosition($newPosition);
+            }
+        });
+    }
+
+    public function setIsSorted(bool $isSorted = true): static
```
Comment (SirSplasch) [inline src/Traits/SortableTrait.php:35]: must this be public or can this be protected? Just curious.

**Review by SirSplasch — CHANGES_REQUESTED** [review]

File: src/Traits/SortableTrait.php (hunk 6 above)
Comment (patrickweh) [reply src/Traits/SortableTrait.php]: can be protected, changed

**Review by SirSplasch — APPROVED** [review]

**Review by SirSplasch — APPROVED** [review]

**Pull request merged**